keep/api/core/dependencies.py

```diff
@@ -206,91 +241,151 @@ def verify_bearer_token_single_tenant(token: str = Depends(oauth2_scheme)) -> st
+            status_code=403,
+            detail="You don't have the required permissions to access this resource",
+        )
+    return AuthenticatedEntity(tenant_id, email)
 
 
-def verify_api_key_single_tenant(
+def _verify_api_key_single_tenant(
     request: Request,
+    scopes: list[str] = [],
     api_key: str = Security(auth_header),
     authorization: HTTPAuthorizationCredentials = Security(http_basic),
     session: Session = Depends(get_session),
-) -> str:
+) -> AuthenticatedEntity:
     # if we don't want to use authentication, return the single tenant id
     if (
         os.environ.get("AUTH_TYPE", AuthenticationType.NO_AUTH.value)
         == AuthenticationType.NO_AUTH.value
     ):
-        return SINGLE_TENANT_UUID
+        return AuthenticatedEntity(
+            tenant_id=SINGLE_TENANT_UUID, email=SINGLE_TENANT_EMAIL
+        )
 
     api_key = __extract_api_key(request, api_key, authorization)
 
     tenant_api_key = get_api_key(api_key)
     if not tenant_api_key:
         raise HTTPException(status_code=401, detail="Invalid API Key")
 
+    role = get_role_by_role_name(tenant_api_key.role)
+    # validate scopes
+    if not role.has_scopes(scopes):
+        raise HTTPException(
+            status_code=403,
+            detail=f"You don't have the required scopes to access this resource [required scoopes: {scopes}]",
+        )
     request.state.tenant_id = tenant_api_key.tenant_id
-    return tenant_api_key.tenant_id
-
-
-def verify_token_or_key(
-    request: Request,
-    api_key: Optional[str] = Security(auth_header),
-    authorization: Optional[HTTPAuthorizationCredentials] = Security(http_basic),
-    token: Optional[str] = Depends(oauth2_scheme),
-) -> str:
-    # Attempt to verify API Key first
-    if api_key:
-        try:
-            return verify_api_key(request, api_key, authorization)
-        except Exception:
-            logger.exception("Failed to validate API Key")
-            raise HTTPException(
-                status_code=401, detail="Invalid authentication credentials"
-            )
-    # If API Key is not present or not valid, attempt to verify the token
-    if token:
-        try:
-            return verify_bearer_token(token)
-        except Exception:
-            logger.exception("Failed to validate token")
-            raise HTTPException(
-                status_code=401, detail="Invalid authentication credentials"
-            )
-    raise HTTPException(status_code=401, detail="Missing authentication credentials")
+    return AuthenticatedEntity(tenant_api_key.tenant_id, tenant_api_key.created_by)
+
+
+def AuthVerifier(scopes: list[str] = []):
+    # Determine the authentication type from the environment variable
+    auth_type = os.environ.get("AUTH_TYPE", AuthenticationType.NO_AUTH.value)
+
+    # Return the appropriate verifier based on the auth type
+    if auth_type == AuthenticationType.MULTI_TENANT.value:
+        return AuthVerifierMultiTenant(scopes)
+    else:
```

Comment:
what happens when it is `NO_AUTH`?

Reply:
its being handled with AuthVerifierSingleTenant (e.g. if